app/src/test/java/org/mozilla/fenix/library/historymetadata/controller/HistoryMetadataGroupControllerTest.kt

```diff
@@ -157,16 +168,68 @@ class HistoryMetadataGroupControllerTest {
     }
 
     @Test
-    fun handleDelete() = testDispatcher.runBlockingTest {
-        controller.handleDelete(setOf(mozillaHistoryMetadataItem, firefoxHistoryMetadataItem))
+    fun handleDeleteSingle() = testDispatcher.runBlockingTest {
+        controller.handleDelete(setOf(mozillaHistoryMetadataItem))
 
         coVerify {
             store.dispatch(HistoryMetadataGroupFragmentAction.Delete(mozillaHistoryMetadataItem))
-            store.dispatch(HistoryMetadataGroupFragmentAction.Delete(firefoxHistoryMetadataItem))
-            historyStorage.deleteHistoryMetadata(mozillaHistoryMetadataItem.historyMetadataKey)
-            historyStorage.deleteHistoryMetadata(firefoxHistoryMetadataItem.historyMetadataKey)
+            historyStorage.deleteVisitsFor(mozillaHistoryMetadataItem.url)
             metrics.track(Event.HistorySearchTermGroupRemoveTab)
         }
+        // Here we don't expect the action to be dispatched, because items inside the store
+        // we provided by getMetadataItemsList(), but only one item has been removed
+        verify(exactly = 0) {
+            browserStore.dispatch(
+                HistoryMetadataAction.DisbandSearchGroupAction(searchTerm = searchTerm)
+            )
+        }
+    }
+
+    @Test
+    fun handleDeleteMultiple() = testDispatcher.runBlockingTest {
+        controller.handleDelete(getMetadataItemsList().toSet())
+
+        coVerify {
+            getMetadataItemsList().forEach {
+                store.dispatch(HistoryMetadataGroupFragmentAction.Delete(it))
+                historyStorage.deleteVisitsFor(it.url)
+            }
+            metrics.track(Event.HistorySearchTermGroupRemoveTab)
+        }
+        // Here we expect the action to be dispatched, because both deleted items and items inside
+        // the store were provided by the same method getMetadataItemsList()
+        verify {
+            browserStore.dispatch(
+                HistoryMetadataAction.DisbandSearchGroupAction(searchTerm = searchTerm)
+            )
+        }
+    }
+
+    @Test
+    fun handleDeleteAbnormal() = testDispatcher.runBlockingTest {
+        val abnormalList = listOf(
+            mozillaHistoryMetadataItem,
+            firefoxHistoryMetadataItem,
+            mozillaHistoryMetadataItem.copy(title = "Pocket", url = "https://getpocket.com"),
+            mozillaHistoryMetadataItem.copy(title = "BBC", url = "https://www.bbc.com/"),
+            mozillaHistoryMetadataItem.copy(title = "Stackoverflow", url = "https://stackoverflow.com/")
+        )
+        controller.handleDelete(abnormalList.toSet())
+        coVerify {
+            getMetadataItemsList().forEach {
```

Comment:
Probably also worth verifying that we're calling `historyStorage.deleteVisitsFor` and `HistoryMetadataGroupFragmentAction.Delete` for the extra items that aren't present in `getMetadataItemsList`.